platform/linux-generic/odp_classification.c

```diff
@@ -1555,56 +1555,48 @@ static inline void pmr_debug_print(pmr_t *pmr, cos_t *cos)
 
 /*
  * Match a PMR chain with a Packet and return matching CoS
- * This function gets called recursively to check the chained PMR Term value
- * with the packet.
+ * This function performs a depth-first search in the CoS tree.
  */
-static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, pmr_t *pmr,
-			    odp_packet_hdr_t *hdr)
+static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, odp_packet_hdr_t *hdr)
 {
-	uint32_t i, num_rule;
+	if (!cos || !cos->valid)
+		return cos;
 
-	if (cos == NULL || pmr == NULL)
-		return NULL;
+	while (1) {
+		uint32_t i, num_rule = odp_atomic_load_u32(&cos->num_rule);
 
-	if (!cos->valid)
-		return NULL;
+		for (i = 0; i < num_rule; i++) {
+			pmr_t *pmr = cos->pmr[i];
 
-	if (verify_pmr(pmr, pkt_addr, hdr)) {
-		/* PMR matched */
-		pmr_debug_print(pmr, cos);
+			if (odp_unlikely(!cos->linked_cos[i]->valid))
+				continue;
 
-		if (cos->stats_enable)
-			odp_atomic_inc_u64(&cos->stats.packets);
+			if (verify_pmr(pmr, pkt_addr, hdr)) {
+				/* PMR matched */
 
-		hdr->p.input_flags.cls_mark = 0;
-		if (pmr->mark) {
-			hdr->p.input_flags.cls_mark = 1;
-			hdr->cls_mark = pmr->mark;
-		}
+				cos = cos->linked_cos[i];
 
-		/* This gets called recursively. First matching leaf or branch
-		 * is returned. */
-		num_rule = odp_atomic_load_u32(&cos->num_rule);
+				pmr_debug_print(pmr, cos);
 
-		/* No more rules. This is the best match. */
-		if (num_rule == 0)
-			return cos;
+				if (cos->stats_enable)
+					odp_atomic_inc_u64(&cos->stats.packets);
 
-		for (i = 0; i < num_rule; i++) {
-			cos_t *retcos = match_pmr_cos(cos->linked_cos[i],
-						      pkt_addr, cos->pmr[i],
-						      hdr);
+				hdr->p.input_flags.cls_mark = 0;
+				if (pmr->mark) {
+					hdr->p.input_flags.cls_mark = 1;
+					hdr->cls_mark = pmr->mark;
```

Comment:
Instead of recording mark into packet (possibly many times), it should be more efficient to just keep record of last matching pmr pointer, and write mark once in the end of the function.